File: shared/shared/permissions.py
"""Permission keys, the built-in roles and their sets, and custom roles (decisions D185, D186).

Endpoints declare the permission they need; the caller's role in the project decides. A role
is a named set of keys: four built-in roles (viewer, operator, analyst, admin) that every
server has, or a custom role a project admin composes from the same keys. A server admin has
every permission in every project. Keys are stable strings so they can appear in the audit
log and in the interface; `AREAS` groups them the way the role editor shows them.
"""

from collections.abc import Iterable
from enum import StrEnum

from shared.enums import Role


class Permission(StrEnum):
    PROJECT_READ = "project:read"
    PROJECT_WRITE = "project:write"
    MEMBERS_WRITE = "members:write"
    ENTITIES_WRITE = "entities:write"
    DEVICES_WRITE = "devices:write"
    FEATURES_WRITE = "features:write"
    DEVICES_CONTROL = "devices:control"
    DEVICES_CONTROL_HIGH_IMPACT = "devices:control_high_impact"
    RULES_WRITE = "rules:write"
    ALERTS_WRITE = "alerts:write"
    EVENTS_WRITE = "events:write"
    AUTOMATIONS_WRITE = "automations:write"
    INTEGRATIONS_WRITE = "integrations:write"
    DATA_CURATE = "data:curate"
    DATA_CURATE_BULK = "data:curate_bulk"
    DATA_APPROVE = "data:approve"
    DATA_REVERT = "data:revert"
    TRACES_READ = "traces:read"
    EXPORTS_CREATE = "exports:create"
    VIEWS_WRITE = "views:write"
    DASHBOARDS_WRITE = "dashboards:write"


#: The keys grouped by area, in the order the role editor shows them. `project:read` is in
#: every role: a member who cannot read the project is not a member.
AREAS: tuple[tuple[str, tuple[Permission, ...]], ...] = (
    ("see", (Permission.PROJECT_READ, Permission.TRACES_READ)),
    (
        "entities_and_devices",
        (Permission.ENTITIES_WRITE, Permission.DEVICES_WRITE, Permission.FEATURES_WRITE),
    ),
    ("events_and_alerts", (Permission.EVENTS_WRITE, Permission.ALERTS_WRITE)),
    (
        "data_quality",
        (
            Permission.DATA_CURATE,
            Permission.DATA_CURATE_BULK,
            Permission.DATA_APPROVE,
            Permission.DATA_REVERT,
        ),
    ),
    ("rules_and_automations", (Permission.RULES_WRITE, Permission.AUTOMATIONS_WRITE)),
    ("integrations", (Permission.INTEGRATIONS_WRITE,)),
    (
        "exports_and_analysis",
        (Permission.EXPORTS_CREATE, Permission.VIEWS_WRITE, Permission.DASHBOARDS_WRITE),
    ),
    ("control", (Permission.DEVICES_CONTROL, Permission.DEVICES_CONTROL_HIGH_IMPACT)),
    ("members_and_settings", (Permission.MEMBERS_WRITE, Permission.PROJECT_WRITE)),
)

ALWAYS = frozenset({Permission.PROJECT_READ})

_VIEWER = frozenset({Permission.PROJECT_READ, Permission.TRACES_READ, Permission.ALERTS_WRITE})
_OPERATOR = _VIEWER | {
    Permission.EVENTS_WRITE,
    Permission.DEVICES_CONTROL,
    Permission.FEATURES_WRITE,
}
_ANALYST = _OPERATOR | {
    Permission.EXPORTS_CREATE,
    Permission.VIEWS_WRITE,
    Permission.DASHBOARDS_WRITE,
}

#: The built-in roles, each a superset of the one before (decision D185).
ROLE_PERMISSIONS: dict[Role, frozenset[Permission]] = {
    Role.PROJECT_VIEWER: _VIEWER,
    Role.PROJECT_OPERATOR: _OPERATOR,
    Role.PROJECT_ANALYST: _ANALYST,
    Role.PROJECT_ADMIN: frozenset(Permission),
}

BUILTIN_ROLE_LABELS: dict[Role, str] = {
    Role.PROJECT_VIEWER: "Viewer",
    Role.PROJECT_OPERATOR: "Operator",
    Role.PROJECT_ANALYST: "Analyst",
    Role.PROJECT_ADMIN: "Admin",
}


def normalise_permissions(keys: Iterable[str]) -> frozenset[Permission]:
    """A custom role's keys as permissions, `project:read` always included; an unknown key
    raises ValueError with the key in the message."""
    result = set(ALWAYS)
    for key in keys:
        try:
            result.add(Permission(key))
        except ValueError:
            raise ValueError(f"unknown permission {key!r}") from None
    return frozenset(result)


def permissions_for(
    role: Role | None,
    *,
    server_admin: bool,
    custom: Iterable[str] | None = None,
) -> frozenset[Permission]:
    """The caller's permissions: everything for a server admin, a custom role's set when the
    membership names one, else the built-in role's set, nothing without a membership."""
    if server_admin:
        return frozenset(Permission)
    if custom is not None:
        return normalise_permissions(custom)
    if role is None:
        return frozenset()
    return ROLE_PERMISSIONS[role]

File: services/frontend/src/lib/permissions.ts
import i18n from "@/i18n";

/**
 * The permission keys as the interface knows them (decision D188): every gate asks `can(key)`
 * on the caller's keys from the project list, and the role editor shows the keys by area with
 * a label and a line of explanation each. The keys and the areas are the API's
 * (`shared/permissions.py`, `GET /permissions`); the words are here so they translate.
 */
export type PermissionKey =
  | "project:read"
  | "project:write"
  | "members:write"
  | "entities:write"
  | "devices:write"
  | "features:write"
  | "devices:control"
  | "devices:control_high_impact"
  | "rules:write"
  | "alerts:write"
  | "events:write"
  | "automations:write"
  | "integrations:write"
  | "data:curate"
  | "data:curate_bulk"
  | "data:approve"
  | "data:revert"
  | "traces:read"
  | "exports:create"
  | "views:write"
  | "dashboards:write"
  | "analysis:run";

export const BUILTIN_ROLES = [
  "project-viewer",
  "project-operator",
  "project-analyst",
  "project-admin",
] as const;
export type BuiltinRole = (typeof BUILTIN_ROLES)[number];

export function roleLabel(role: string): string {
  const t = i18n.t.bind(i18n);
  switch (role) {
    case "project-viewer":
      return t("Viewer");
    case "project-operator":
      return t("Operator");
    case "project-analyst":
      return t("Analyst");
    case "project-admin":
      return t("Admin");
    case "server-admin":
      return t("Server admin");
    default:
      return role;
  }
}

export function roleDescription(role: string): string {
  const t = i18n.t.bind(i18n);
  switch (role) {
    case "project-viewer":
      return t("Sees the project, reads traces, acknowledges alerts.");
    case "project-operator":
      return t("A viewer who also reports events, controls devices and draws features.");
    case "project-analyst":
      return t("An operator who also exports data and keeps saved views and dashboards.");
    case "project-admin":
      return t("Everything in the project, members and settings included.");
    default:
      return "";
  }
}

export function areaLabel(area: string): string {
  const t = i18n.t.bind(i18n);
  const labels: Record<string, string> = {
    see: t("See"),
    entities_and_devices: t("Entities and devices"),
    events_and_alerts: t("Events and alerts"),
    data_quality: t("Data quality"),
    rules_and_automations: t("Rules and automations"),
    integrations: t("Integrations"),
    exports_and_analysis: t("Exports and analysis"),
    control: t("Control"),
    members_and_settings: t("Members and settings"),
  };
  return labels[area] ?? area;
}

export function permissionLabel(key: string): string {
  const t = i18n.t.bind(i18n);
  const labels: Record<string, string> = {
    "project:read": t("See the project"),
    "traces:read": t("Read processing traces"),
    "entities:write": t("Create and edit entities and groups"),
    "devices:write": t("Assign devices"),
    "features:write": t("Draw and edit features"),
    "events:write": t("Report events"),
    "alerts:write": t("Acknowledge and resolve alerts"),
    "data:curate": t("Curate records"),
    "data:curate_bulk": t("Curate in bulk"),
    "data:approve": t("Approve curation"),
    "data:revert": t("Revert curation"),
    "rules:write": t("Edit rules"),
    "automations:write": t("Edit automations"),
    "integrations:write": t("Edit integrations"),
    "exports:create": t("Export data"),
    "views:write": t("Keep saved views"),
    "dashboards:write": t("Build dashboards"),
    "analysis:run": t("Run analyses"),
    "devices:control": t("Send commands to devices"),
    "devices:control_high_impact": t("Send high-impact commands"),
    "members:write": t("Manage members and roles"),
    "project:write": t("Change project settings"),
  };
  return labels[key] ?? key;
}

export function permissionHint(key: string): string {
  const t = i18n.t.bind(i18n);
  const hints: Record<string, string> = {
    "project:read": t("Always on: a member who cannot see the project is not a member."),
    "traces:read": t("The Trace explorer and the trace of a record."),
    "entities:write": t("Entities, their pictures and the group tree."),
    "devices:write": t("Assign a device to an entity or release it."),
    "features:write": t("Sites, zones, geofences and routes on the map."),
    "events:write": t("Manual events, such as a sighting."),
    "alerts:write": t("Take ownership of an alert and close it."),
    "data:curate": t("Mark records invalid or correct them one by one."),
    "data:curate_bulk": t("Curation jobs over many records."),
    "data:approve": t("Approve proposed curation."),
    "data:revert": t("Undo curation."),
    "rules:write": t("Rules that create events and alerts."),
    "automations:write": t("What happens after an event."),
    "integrations:write": t("Deliveries to other platforms."),
    "exports:create": t("Export jobs and direct downloads."),
    "views:write": t("Saved views on the explore page."),
    "dashboards:write": t("Dashboards of the project."),
    "analysis:run": t("Movement and grazing analyses and their kept results."),
    "devices:control": t("Status and position requests and settings."),
    "devices:control_high_impact": t("Commands that change how a device behaves for good."),
    "members:write": t("Invite people, set roles and scopes, compose roles."),
    "project:write": t("The project's name, settings, icons and notifications."),
  };
  return hints[key] ?? "";
}
